apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/helper/RecordingServletInputStreamWrapper.java

```diff
@@ -0,0 +1,190 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.servlet.helper;
+
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.impl.context.Request;
+import co.elastic.apm.agent.util.IOUtils;
+
+import javax.servlet.ReadListener;
+import javax.servlet.ServletInputStream;
+import java.io.IOException;
+import java.nio.CharBuffer;
+import java.nio.charset.CoderResult;
+
+@VisibleForAdvice
+public class RecordingServletInputStreamWrapper extends ServletInputStream {
+
+    private final Request request;
+    private final ServletInputStream servletInputStream;
+
+    @VisibleForAdvice
+    public RecordingServletInputStreamWrapper(Request request, ServletInputStream servletInputStream) {
+        this.request = request;
+        this.servletInputStream = servletInputStream;
+    }
+
+    @Override
+    public int read() throws IOException {
+        try {
```

Comment:
Can we only instrument this one? From a quick look it seems all others are calling this one.

Reply:
We can't because the methods of `ServletInputStream` will very likely be overridden by the actual implementation.